packages/app/src/cli/services/dev/urls.ts

```diff
@@ -86,23 +79,21 @@ export async function generateFrontendURL(options: FrontendURLOptions): Promise<
     return {frontendUrl, frontendPort, usingLocalhost}
   }
 
-  if (needsTunnel) {
-    frontendPort = await getAvailableTCPPort()
-    frontendUrl = await generateURL(options.commandConfig, frontendPort)
-  } else {
+  if (options.noTunnel) {
     frontendPort = await getAvailableTCPPort()
     frontendUrl = 'http://localhost'
     usingLocalhost = true
+  } else {
+    frontendPort = await getAvailableTCPPort()
+    const provider = options.tunnelProvider || options.useCloudflareTunnels ? 'cloudflare' : 'ngrok'
```

Comment:
The second expression is missing parentheses, so the logic is not working as expected

```suggestion
    const defaultProvider = options.useCloudflareTunnels ? 'cloudflare' : 'ngrok'
    const provider = options.tunnelProvider || defaultProvider
```